source/agora/node/admin/QRCodeInterface.d

```diff
@@ -0,0 +1,289 @@
+
+        auto app = req.query["app"];
+        auto height = Height.fromString(req.query["height"]);
+
+        // TODO: PreImage to be released at that height
+        // Or it can be pre-released, so it can be `height + PreimageRevealPeriod`
+        PreImageInfo pre_image;
+
+        // blake2b hashing
+        const Hash key = hashMulti(pre_image, app);
+
+        EncryptionKey encryptionKey = EncryptionKey(
+            app,
+            height,
+            key
+        );
+
+        res.headers["Content-Type"] = "image/svg+xml";
+        res.headers["Vary"] = "Accept-Encoding";
+        res.writeBody(createQRcode(encryptionKey));
+    }
+
+    /***************************************************************************
+
+        SVG format QR code generator
+
+        Use barcode library to make QR code of JSON format.
+        Convert the data structure to JSON and return the QR code image
+        of the SVG format.
+
+        Params:
+            t = Type of struct to QR code
+
+        Returns:
+            QR code of XML string
+
+    ***************************************************************************/
+
+    private string createQRcode (T)(T t)
+        if (is(T == struct))
+    {
+        auto str = serializeToJsonString(t);
+
+        auto svgDrawer = new BaseBarCodeSvgDrawer;
+        svgDrawer.fixSizeMode = true;
+        svgDrawer.W = 400;
+        svgDrawer.H = 400;
+
+        auto qr = new Qr;
+        auto bc = qr.encode(str);
+        return(svgDrawer.draw(bc));
+    }
+}
+
+/*******************************************************************************
+
+    Defines the LoginInfo.
+
+    This includes the user's credentials.
+    Includes a temporary private key used to prevent leakage of the
+    validator's private key.
+
+*******************************************************************************/
+
+public struct LoginInfo
+{
+    // The randomly generated temporary private key encoded as string
+    public string private_key;
+
+    // The `Voter card`
+    public VoterCard voter_card;
+}
+
+/*******************************************************************************
+
+    Defines the VoterCard.
+
+    Voter card provided at login stage
+    The voter card contains an expiration date set at 90 days
+    from the time of creation.
+    The signature is signed with the validator's private key
+
+*******************************************************************************/
+
+public struct VoterCard
+{
+    // Validator that this voter card represents
+    public PublicKey validator;
+
+    // Public key of the `private_key` field
+    public PublicKey address;
+
+    // A 'nonce' used to allow expiring / replacing credentials
+    // In this case, we just use an expiration date in ISO8601
+    public string expires;
+
+    // Finally, the signature, made using `validator_address` private key
+    // (IOW: the `Validator` private key) on this whole packet.
+    public Hash signature;
```

Comment:
Why is this a `Hash` instead of a `Signature`?

Reply:
I declared it hash for compatibility with boa-sdk.
Signature is BitBlob!512 is an alias, so I will modify it to Signature .